### DISCUSSION TITLE
Support for urls in magic `$`-imports

### DISCUSSION THREAD

**Discussion by lefou** [com-lihaoyi/mill#2996] [Ideas]
The `$file` import currently accepts a source file, compiles it with the current Mill project classpath and adds it to the classpath.

It would be nice, if we could have a variant that accepts remote files. It should work the same, just that the file is downloaded locally before it is further processed.

Futhermore, it should also support `.scala` files, not just `.sc` files. There reason is, that `.sc` files need to be surrounded with a synthetic object. Once inside such sythetic object, it' no longer possible to contribute external commands to the build.

Once we have that support, we can also support it via the `--import` cli option.

And here is my motivation:

I'd like to provide rather simple plugins as external modules, and a simple Scala source file would be the easiest solution for that. For example: https://github.com/lefou/millw/issues/40

Imagine a simple remote plugin provided as a source file.

```
> mill --import url:https://gist.github.com/my/file my.file.Module/command
```

**Comment by lihaoyi** [comment]
We had this feature in Ammonite https://ammonite.io/#1.0.2

I didn't port it to Mill because I never really used it, but could be added if we really wanted

One concern is that it breaks the hermeticity of your build. But if you're importing stuff from random URLs, you probably know what you're getting yourself into, and hopefully know what you're doing

**Reply by lefou** [reply]
My use-case is not so much the core build, but rather some random code analyzer, formatter, exporter tool typically used as external module. Hence, I think it would be necessary to only start with `--import` which adds runtime-only-dependencies to Mill.
